packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts

```
@@ -5,21 +5,145 @@
  * Use of this source code is governed by an MIT-style license that can be
  * found in the LICENSE file at https://angular.io/license
  */
-import {ConstantPool} from '@angular/compiler';
+import {compileDirectiveFromMetadata, ConstantPool, makeBindingParser, ParseLocation, ParseSourceFile, ParseSourceSpan, R3DirectiveMetadata, R3HostMetadata, R3QueryMetadata, R3Reference} from '@angular/compiler';
 import * as o from '@angular/compiler/src/output/output_ast';
 
-import {AstObject} from '../../ast/ast_value';
+import {Range} from '../../ast/ast_host';
+import {AstObject, AstValue} from '../../ast/ast_value';
+import {FatalLinkerError} from '../../fatal_linker_error';
 
 import {PartialLinker} from './partial_linker';
 
 /**
  * A `PartialLinker` that is designed to process `$ngDeclareDirective()` call expressions.
  */
-export class PartialDirectiveLinkerVersion1<TStatement, TExpression> implements
-    PartialLinker<TStatement, TExpression> {
+export class PartialDirectiveLinkerVersion1<TExpression> implements PartialLinker<TExpression> {
   linkPartialDeclaration(
       sourceUrl: string, code: string, constantPool: ConstantPool,
       metaObj: AstObject<TExpression>): o.Expression {
-    throw new Error('Not implemented.');
+    const meta = toR3DirectiveMeta(metaObj, code, sourceUrl);
+    const def = compileDirectiveFromMetadata(meta, constantPool, makeBindingParser());
+    return def.expression;
   }
 }
+
+/**
+ * Derives the `R3DirectiveMetadata` structure from the AST object.
+ */
+export function toR3DirectiveMeta<TExpression>(
+    metaObj: AstObject<TExpression>, code: string, sourceUrl: string): R3DirectiveMetadata {
+  const typeName = metaObj.getValue('type').getSymbolName() ?? 'anonymous';
```

Is "anonymous" a valid scenario? Perhaps we should throw an error here?
From digging through the code, it looks like this directive type name is used to create names for queries and template functions. I think it would be worth, if not erroring, making sure that this is unique for the given constant pool?

---

The `type` property is not optional on the `R3DeclareDirectiveMetadata` interface, so it seems unlikely (impossible?) that a directive would be defined as an anonymous class?

---

Yes I agree this case should throw instead of the magic anonymous name.